src/js/service/VendingMachine.js

```diff
@@ -1,67 +1,97 @@
 
 export class VendingMachine {
-  #products;
+  /** @type {ProductManager} */
+  #productManager;
   /**@type {UnitCountMachine} */
   #unitCountMachine;
+  /**@type {number} */
+  #insertedMoney;
 
-  constructor() {
-    this.reset();
-  }
-
-  reset() {
-    this.#products = [];
-    this.#unitCountMachine = new UnitCountMachine();
+  /**
+   * @param {?ProductItem[]} products
+   * @param {?UnitCountInfo} unitCountInfo
+   */
+  constructor(products, unitCountInfo) {
+    this.#productManager = new ProductManager(products);
+    this.#unitCountMachine = new UnitCountMachine(unitCountInfo);
+    this.#insertedMoney = 0;
   }
 
   /**
    *
-   * @returns {VendingMachineItem[]}
+   * @returns {ProductManager}
    */
-  get products() {
-    return this.#products;
+  get productManager() {
+    return this.#productManager;
   }
 
+  /**
+   * @returns {UnitCountMachine}
+   */
   get unitCountMachine() {
     return this.#unitCountMachine;
   }
 
-  /**
-   *
-   * @param {VendingMachineItem} vendingMachineItem
-   */
-  addItem({ name, price, amount }) {
-    this.#validateProduct({ name, price, amount });
-    this.#products = [
-      ...this.#products.filter((item) => item.name !== name),
-      {
-        index: this.#products,
-        name: removeSpace(name),
-        price,
-        amount,
-      },
-    ].sort((a, b) => a.index - b.index);
+  get insertedMoney() {
+    return this.#insertedMoney;
   }
 
   /**
-   *
-   * @param {VendingMachineItem} vendingMachineItem
+   * @param {number} index
+   * @returns {ProductItem}
    */
-  #validateProduct({ name, price, amount }) {
-    if (!isNameValid(name)) {
-      throw new ValidationError(ALERT_MESSAGE.VALIDATION.PRODUCT.NAME_BLANK);
+  purchase(index) {
+    const product = this.#productManager.products[index];
+    if (!(product && this.#insertedMoney - product.price >= 0)) {
+      throw new ValidationError(ERROR_MESSAGE.NOT_ENOUGH_SPENDING_MONEY);
     }
-    if (!isPriceValid(price)) {
-      throw new ValidationError(ALERT_MESSAGE.VALIDATION.PRODUCT.PRICE);
+    if (product.amount === 0) {
+      throw new ValidationError(ERROR_MESSAGE.SOLD_OUT);
     }
-    if (!isAmountValid(amount)) {
-      throw new ValidationError(ALERT_MESSAGE.VALIDATION.PRODUCT.AMOUNT);
+
+    this.#insertedMoney -= product.price;
+    this.#productManager.redraw(index);
+    return product;
+  }
+
+  /**
+   * @param {number} amount
+   */
+  insertMoney(amount) {
+    if (!VendingMachine.#isInsertedMoneyValid(amount)) {
+      throw new ValidationError(ERROR_MESSAGE.VALIDATION.SPENDING_MONEY_INPUT);
     }
+    this.#insertedMoney += amount;
   }
+
+  returnChanges() {
+    const { amount, unitInfo } = this.#unitCountMachine.redraw(this.#insertedMoney);
+    this.#insertedMoney -= amount;
+    return { amount, unitInfo };
+  }
+
+  static #isInsertedMoneyValid = (amount) =>
+    isInteger(amount) && isGreaterThan(amount, MIN_AMOUNT) && isMultipleOf(amount, MULTIPLE);
```

Comment:
```suggestion
  insertMoney(amount) {
    !VendingMachine.#validateInsertedMoney(amount));
    this.#insertedMoney += amount;
  }

  returnChanges() {
    const { amount, unitInfo } = this.#unitCountMachine.redraw(this.#insertedMoney);
    this.#insertedMoney -= amount;
    return { amount, unitInfo };
  }

  static #validateInsertedMoney(amount) {
    if(isInteger(amount) && isGreaterThan(amount, MIN_AMOUNT) && isMultipleOf(amount, MULTIPLE)) {
      throw new ValidationError(ERROR_MESSAGE.VALIDATION.SPENDING_MONEY_INPUT)
    }
  }
```

현재 상태는 검증에 대한 책임을 insertMoeny와 isInsertedMoneyValid가 나눠가진 상태입니다.
이걸 제시한 것 처럼 아예 한 곳에 몰아주면 어떨까요!?